x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/mapreduce/MapReduceAggregator.java

```diff
@@ -0,0 +1,176 @@
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.aggs.mapreduce;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.util.SetOnce;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.common.util.LongObjectPagedHashMap;
+import org.elasticsearch.core.Releasables;
+import org.elasticsearch.search.aggregations.Aggregator;
+import org.elasticsearch.search.aggregations.AggregatorBase;
+import org.elasticsearch.search.aggregations.AggregatorFactories;
+import org.elasticsearch.search.aggregations.CardinalityUpperBound;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.LeafBucketCollector;
+import org.elasticsearch.search.aggregations.LeafBucketCollectorBase;
+import org.elasticsearch.search.aggregations.support.AggregationContext;
+import org.elasticsearch.search.aggregations.support.ValuesSourceConfig;
+import org.elasticsearch.xcontent.ToXContent;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+
+import static org.elasticsearch.xpack.ml.aggs.mapreduce.MapReduceValueSourceRegistry.REGISTRY_KEY;
+import static org.elasticsearch.xpack.ml.aggs.mapreduce.MapReduceValueSourceRegistry.SUPPORTED_TYPES;
+
+public abstract class MapReduceAggregator<
+    MapContext extends Closeable,
+    MapFinalContext extends Writeable,
+    ReduceContext extends Closeable,
+    Result extends ToXContent & Writeable> extends AggregatorBase {
+
+    private final List<ValuesExtractor> extractors;
+    private final List<String> fieldNames;
+    private final AbstractMapReducer<MapContext, MapFinalContext, ReduceContext, Result> mapReducer;
+    private final BigArrays bigArraysForMapReduce;
+    private final LongObjectPagedHashMap<Object> mapReduceContextByBucketOrdinal;
+    private final boolean profiling;
+    private final DelegatingCircuitBreakerService breakerService;
+
+    protected MapReduceAggregator(
+        String name,
+        AggregationContext context,
+        Aggregator parent,
+        Map<String, Object> metadata,
+        AbstractMapReducer<MapContext, MapFinalContext, ReduceContext, Result> mapReducer,
+        List<ValuesSourceConfig> configs
+    ) throws IOException {
+        super(name, AggregatorFactories.EMPTY, context, parent, CardinalityUpperBound.NONE, metadata);
+
+        List<ValuesExtractor> extractors = new ArrayList<>();
+        List<String> fieldNames = new ArrayList<>();
+        int id = 0;
+        for (ValuesSourceConfig c : configs) {
+            if (SUPPORTED_TYPES.contains(c.valueSourceType()) == false) {
+                throw new IllegalArgumentException(
+                    c.getDescription() + " is not supported for aggregation [" + mapReducer.getWriteableName() + "]"
```

Comment:
This should be unnecessary.  The `ValuesSourceRegistry` will throw the correct exception when you call `getAggregator` with an unmapped values source type.

(sorry for the duplicate comment, I replied to the commit first, which apparently doesn't work right)

Reply:
its now in the agg, so this override has been removed